### DISCUSSION TITLE
Long capture lists

### DISCUSSION THREAD

**Discussion by dcramps** [airbnb/swift#188] [General]
Consider some code dealing with `Operations` and `BlockOperations` that requires the use of a capture list. Generally, capture lists are just `[weak self]`, but in this case they're a bit more verbose:

```
let someOperation = SomeOperation(parameter: object)
let someOtherOperation = SomeOtherOperation(parameter: object)
let thirdOperation = ThirdOperation(parameter: object)

let blockOperation = BlockOperation { [unowned someOperation, unowned someOtherOperation, unowned thirdOperation] in
  // do a thing
}

...
```

This clearly violates the 100 character line rule, so I looked at a few ways of solving:

```
let someOperation = SomeOperation(parameter: object)
let someOtherOperation = SomeOtherOperation(parameter: object)
let thirdOperation = ThirdOperation(parameter: object)

let blockOperation = BlockOperation { 
  [unowned someOperation, unowned someOtherOperation, unowned thirdOperation] in
  // do a thing
}

...
```
Problem: the capture list is at the same level of indentation as the code in the block

```
let someOperation = SomeOperation(parameter: object)
let someOtherOperation = SomeOtherOperation(parameter: object)
let thirdOperation = ThirdOperation(parameter: object)

let blockOperation = BlockOperation 
{ [unowned someOperation, unowned someOtherOperation, unowned thirdOperation] in
  // do a thing
}

...
```
Problem: same as above, plus the line is still too long

```
let someOperation = SomeOperation(parameter: object)
let someOtherOperation = SomeOtherOperation(parameter: object)
let thirdOperation = ThirdOperation(parameter: object)

let blockOperation = BlockOperation 
{ 
  [
  unowned someOperation, 
  unowned someOtherOperation, 
  unowned thirdOperation
  ] 
  in

  // do a thing
}

...
```

Problem: same as above, but at least there's a space between `in` and the code to be executed. This looks really weird too. Capture lists don't indent like arrays, so the opening and closing brackets are at the same level of indentation as the contents of the capture list.

There doesn't seem to be a way to split this into multiple lines cleanly without having the capture list on the same level of indentation as the code, so we need to come up with a "good enough" solution that clearly separates the capture list from the code.

Any thoughts? I'm leaning towards a mix of some of these ideas, but still not super happy with it:

```
let someOperation = SomeOperation(parameter: object)
let someOtherOperation = SomeOtherOperation(parameter: object)
let thirdOperation = ThirdOperation(parameter: object)

let blockOperation = BlockOperation 
{ 
  [unowned someOperation, unowned someOtherOperation, unowned thirdOperation] 
  in

  // do a thing
}

...
```

@airbnb/swift-styleguide-maintainers

**Comment by bachand** [comment]
This is a tough one.

What do we do in the last example if the capture list exceeds 100 characters?

**Comment by dimitryz-luxuryretreats** [comment]
I personally have not yet seen a situation where there are that many arguments in a single line, so I don't really mind that a line goes beyond the 100 char limit every now and again. However, if there had to be a suggestion of how to format this line, I would say something like this:

```
let someOperation = SomeOperation(parameter: object)
let someOtherOperation = SomeOtherOperation(parameter: object)
let thirdOperation = ThirdOperation(parameter: object)

let blockOperation = BlockOperation {
  [unowned someOperation, unowned someOtherOperation, unowned thirdOperation] in
  // do a thing
}

...
```

**Comment by dcramps** [comment]
@bachand I guess in that case we would do what we do with arrays, and live with the oddness of it?
@dimitryz-luxuryretreats what do you think about an empty line between the list and the code to make a clear separation?

**Comment by dimitryz-luxuryretreats** [comment]
Yeah, that makes sense too. We usually put an empty line after a guard statement too:

```
view.action = { [weak self] in
  guard let self = self else { return }

  // code
}
```

**Comment by bachand** [comment]
I think whatever we come up with should handle long parameter lists too:
<img width="1121" alt="screen shot 2018-09-27 at 1 19 26 am" src="https://user-images.githubusercontent.com/1791049/46124652-645f9700-c1f3-11e8-8030-f2f302e1fcd2.png">

**Comment by bachand** [comment]
In this case we can convert to `$0` and `$1` but that won't always be true.
